### PR TITLE
feat: fast and standard queue in CES controller

### PR DESCRIPTION
**Pull request by Kaczyniec**
In order to allow to enforce Network Policy quicker in more critical namespaces, we propose to introduce a second queue in the CES controller. 
CES and CEP updates from namespaces listed as prioritized in the controller's attribute priorityNamespaces are added to fast queue and processed first. The namespaces are listed in the priorityNamespaces attribute if they have an appropriate annotation.
[Feature proposal doc](https://docs.google.com/document/d/13yYuMYrmhQB1ppnM9oYVO9yq5ga6qN2GBTlU4lFaPD4/edit?usp=sharing)

### CODE REVIEW

File: operator/pkg/ciliumendpointslice/controller.go [hunk 1]
```diff
@@ -63,15 +69,25 @@ type Controller struct {
...
 	wp *workerpool.WorkerPool
 
 	metrics *Metrics
+
+	defaultCESSyncTime time.Duration
+
+	priorityNamespaces map[string]int
+
+	cond     sync.Cond
+	condLock sync.Mutex
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/controller.go]: remove `condLock` and initialise `cond` in-place

File: operator/pkg/ciliumendpointslice/endpointslice.go [hunk 2]
```diff
@@ -59,30 +59,61 @@ func (c *Controller) initializeQueue() {
...
+func (c *Controller) addToQueue(ces CESName) {
+	_, ok := c.priorityNamespaces[ces.Namespace]
+
+	if ok {
+		time.AfterFunc(c.defaultCESSyncTime, func() {
+			c.logger.Infof("CES %s, %s added to the fast queue after %d", ces.Name, ces.Namespace, c.defaultCESSyncTime)
+			c.fastQueue.Add(ces)
+			c.cond.Signal()
+		})
+		//c.fastQueue.AddAfter(ces, c.defaultCESSyncTime)
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/endpointslice.go]: remove comments

File: operator/pkg/ciliumendpointslice/controller.go [hunk 3]
```diff
@@ -63,15 +69,25 @@ type Controller struct {
...
+
+	rateLimiter workqueue.RateLimiter
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/controller.go]: can we group rateLimiter with rateLimitConfig

File: operator/pkg/ciliumendpointslice/controller.go [hunk 4]
```diff
@@ -63,15 +69,25 @@ type Controller struct {
 	// CES requests going to api-server, ensures a single CES will not be proccessed
 	// multiple times concurrently, and if CES is added multiple times before it
 	// can be processed, this will only be processed only once.
-	queue     workqueue.RateLimitingInterface
-	rateLimit rateLimitConfig
+	fastQueue     workqueue.RateLimitingInterface
+	standardQueue workqueue.RateLimitingInterface
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/controller.go]: Can we add a short description how this works

File: operator/pkg/ciliumendpointslice/endpointslice.go [hunk 5]
```diff
@@ -59,30 +59,61 @@ func (c *Controller) initializeQueue() {
 		logfields.WorkQueueBurstLimit:  c.rateLimit.current.Burst,
 		logfields.WorkQueueSyncBackOff: defaultSyncBackOff,
 	}).Info("CES controller workqueue configuration")
-	c.queue = workqueue.NewRateLimitingQueueWithConfig(
-		workqueue.NewItemExponentialFailureRateLimiter(defaultSyncBackOff, maxSyncBackOff),
+	c.rateLimiter = workqueue.NewItemExponentialFailureRateLimiter(defaultSyncBackOff, maxSyncBackOff)
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/endpointslice.go]: add a small comment explaining why we can have only one rateLimiter as the keys should not be present in both queues

File: operator/pkg/ciliumendpointslice/endpointslice.go [hunk 6]
```diff
@@ -59,30 +59,61 @@ func (c *Controller) initializeQueue() {
...
 func (c *Controller) onSliceDelete(ces *capi_v2a1.CiliumEndpointSlice) {
-	c.enqueueCESReconciliation([]CESName{NewCESName(ces.Name)})
+	c.logger.Infof("on CES delete %s, %s", ces.Name, ces.Namespace)
+	c.enqueueCESReconciliation([]CESName{NewCESNameNamespace(ces.Name, ces.Namespace)})
+}
+
+func (c *Controller) addToQueue(ces CESName) {
+	_, ok := c.priorityNamespaces[ces.Namespace]
+
+	if ok {
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/endpointslice.go]:
```suggestion
	if _, exists := c.priorityNamespaces[ces.Namespace]; exists {
```

File: operator/pkg/ciliumendpointslice/controller.go [hunk 7]
```diff
@@ -63,15 +69,25 @@ type Controller struct {
...
 
 	wp *workerpool.WorkerPool
 
 	metrics *Metrics
+
+	defaultCESSyncTime time.Duration
+
+	priorityNamespaces map[string]int
+
+	cond     sync.Cond
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/controller.go]: can we add a comment to the purpose of `cond`

File: operator/pkg/ciliumendpointslice/endpointslice.go [hunk 8]
```diff
@@ -59,30 +59,61 @@ func (c *Controller) initializeQueue() {
...
+		time.AfterFunc(c.defaultCESSyncTime, func() {
+			c.logger.Infof("CES %s, %s added to the fast queue after %d", ces.Name, ces.Namespace, c.defaultCESSyncTime)
+			c.fastQueue.Add(ces)
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/endpointslice.go:103]: we could potentially check the `if exists` condition here instead of having two time.AfterFunc

File: operator/pkg/ciliumendpointslice/endpointslice.go [hunk 9]
```diff
@@ -252,17 +284,41 @@ func (c *Controller) rateLimitProcessing() {
 	}
 }
 
+func (c *Controller) getQueue() workqueue.RateLimitingInterface {
+	var queue workqueue.RateLimitingInterface
+	c.cond.L.Lock()
+	defer c.cond.L.Unlock()
+	c.logger.Infof("Standard queue length: %d, fast queue length: %d", c.standardQueue.Len(), c.fastQueue.Len())
+	if c.fastQueue.Len() == 0 && c.standardQueue.Len() == 0 {
+		c.logger.Infof("empty queues")
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/endpointslice.go]: logs are too verbose

File: operator/pkg/ciliumendpointslice/endpointslice.go [hunk 10]
```diff
@@ -252,17 +284,41 @@ func (c *Controller) rateLimitProcessing() {
 	}
 }
 
+func (c *Controller) getQueue() workqueue.RateLimitingInterface {
+	var queue workqueue.RateLimitingInterface
+	c.cond.L.Lock()
+	defer c.cond.L.Unlock()
+	c.logger.Infof("Standard queue length: %d, fast queue length: %d", c.standardQueue.Len(), c.fastQueue.Len())
+	if c.fastQueue.Len() == 0 && c.standardQueue.Len() == 0 {
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/endpointslice.go:308]: we might need a for-loop here to block this as long as the queues are empty.

File: operator/pkg/ciliumendpointslice/endpointslice.go [hunk 11]
```diff
@@ -252,17 +284,41 @@ func (c *Controller) rateLimitProcessing() {
...
+		c.logger.Infof("empty queues")
+		c.cond.Wait()
+		c.logger.Infof("End of waiting")
+	}
+
+	if c.fastQueue.Len() == 0 {
+		queue = c.standardQueue
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/endpointslice.go]: we can just return here the queue

File: operator/pkg/ciliumendpointslice/endpointslice.go [hunk 12]
```diff
@@ -252,17 +284,41 @@ func (c *Controller) rateLimitProcessing() {
...
 		return false
 	}
 	key := cKey.(CESName)
+	defer queue.Done(key)
+
 	c.logger.WithFields(logrus.Fields{
 		logfields.CESName: key.string(),
 	}).Debug("Processing CES")
-	defer c.queue.Done(key)
+	defer queue.Done(key)
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/endpointslice.go]: you already called Done on line 316

**Review by ovidiutirla — CHANGES_REQUESTED** [review]
had a quick look over. Please reduce the logging verbosity.

Other than that looks good 🚀 Thanks!

File: operator/pkg/ciliumendpointslice/namespace.go [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+	k8s "github.com/cilium/cilium/pkg/annotation"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slimcorev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+func (c *Controller) processNamespaceEvents(ctx context.Context) error {
+
+	//store, _ := c.namespace.Store(c.context)
```
Comment (dlapcevic) [inline operator/pkg/ciliumendpointslice/namespace.go]: Please remove the commented code before sending the PR for review

File: operator/pkg/ciliumendpointslice/endpointslice_test.go [hunk 14]
```diff
@@ -151,13 +156,195 @@ func TestIdentityModeSyncCESsInLocalCache(t *testing.T) {
...
+		assert.Equal(t, 1, cesController.standardQueue.Len())
+	}
+	cesController.processNextWorkItem()
+	//Enqueue the same CEP (and CES) to test if reconciler catches the repetition
+	cesController.onEndpointUpdate(cep1)
+
+	queue := cesController.getQueue()
+	cKey, _ := queue.Get()
+	key := cKey.(CESName)
+	if err := testutils.WaitUntil(func() bool {
```
Comment (dlapcevic) [inline operator/pkg/ciliumendpointslice/endpointslice_test.go:337]: Please add a comment why this is checked.

Only key.Name is part of the CES's key in the cesStore.

File: operator/pkg/ciliumendpointslice/controller.go [hunk 15]
```diff
@@ -97,14 +115,24 @@ func registerController(p params) error {
...
 		rateLimit:           rateLimitConfig,
 		enqueuedAt:          make(map[CESName]time.Time),
 		metrics:             p.Metrics,
+		defaultCESSyncTime:  DefaultCESSyncTime,
+		priorityNamespaces:  make(map[string]struct{}),
+		condLock:            lock.Mutex{},
+		Job:                 p.Job,
 	}
-
+	cesController.cond = *sync.NewCond(&cesController.condLock)
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/controller.go]:
```suggestion
	cesController.cond = *sync.NewCond(&lock.Mutex{})
```

File: operator/pkg/ciliumendpointslice/controller.go [hunk 16]
```diff
@@ -97,14 +115,24 @@ func registerController(p params) error {
...
 		ciliumNodes:         p.CiliumNodes,
+		namespace:           p.Namespace,
 		slicingMode:         p.Cfg.CESSlicingMode,
 		maxCEPsInCES:        p.Cfg.CESMaxCEPsInCES,
 		rateLimit:           rateLimitConfig,
 		enqueuedAt:          make(map[CESName]time.Time),
 		metrics:             p.Metrics,
+		defaultCESSyncTime:  DefaultCESSyncTime,
+		priorityNamespaces:  make(map[string]struct{}),
+		condLock:            lock.Mutex{},
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/controller.go]:
```suggestion
```

File: operator/pkg/ciliumendpointslice/controller.go [hunk 17]
```diff
@@ -63,15 +69,27 @@ type Controller struct {
...
 	wp *workerpool.WorkerPool
 
 	metrics *Metrics
+
+	defaultCESSyncTime time.Duration
+
+	priorityNamespaces map[string]struct{}
+
+	cond     sync.Cond
+	condLock lock.Mutex
```
Comment (ovidiutirla) [inline operator/pkg/ciliumendpointslice/controller.go]: We could potentially remove this condLock and initilize in-line with cond? see diff below
```suggestion
```

File: operator/pkg/ciliumendpointslice/endpointslice.go (hunk 10 above)
Comment (Kaczyniec) [reply operator/pkg/ciliumendpointslice/endpointslice.go:308]: Added Wait until condition to solve this

File: operator/pkg/ciliumendpointslice/controller.go [hunk 18]
```diff
@@ -63,15 +69,30 @@ type Controller struct {
 	// CES requests going to api-server, ensures a single CES will not be proccessed
 	// multiple times concurrently, and if CES is added multiple times before it
 	// can be processed, this will only be processed only once.
-	queue     workqueue.RateLimitingInterface
-	rateLimit rateLimitConfig
+	// Updates from CEP and CES in namespaces annotated as priority are added to the
+	// first queue and processed first in order to ensure faster enforcement of the
```
Comment (dlapcevic) [inline operator/pkg/ciliumendpointslice/controller.go]: nit: "are added to the ~~first~~ fast queue"

File: operator/pkg/ciliumendpointslice/controller.go [hunk 19]
```diff
@@ -63,15 +69,30 @@ type Controller struct {
...
 
 	wp *workerpool.WorkerPool
 
 	metrics *Metrics
+
+	defaultCESSyncTime time.Duration
+
+	priorityNamespaces map[string]struct{}
+
+	// if the queues are empty, they wait until the condition (adding something to the queues) is met
```
Comment (dlapcevic) [inline operator/pkg/ciliumendpointslice/controller.go]: nit: Format the comment (first letter capital and end the comment with a period)

File: operator/pkg/ciliumendpointslice/controller.go [hunk 20]
```diff
@@ -97,14 +118,23 @@ func registerController(p params) error {
...
 		maxCEPsInCES:        p.Cfg.CESMaxCEPsInCES,
 		rateLimit:           rateLimitConfig,
 		enqueuedAt:          make(map[CESName]time.Time),
 		metrics:             p.Metrics,
+		defaultCESSyncTime:  DefaultCESSyncTime,
+		priorityNamespaces:  make(map[string]struct{}),
+		Job:                 p.Job,
 	}
-
+	cesController.cond = *sync.NewCond(&lock.Mutex{})
```
Comment (dlapcevic) [inline operator/pkg/ciliumendpointslice/controller.go]: nit: `cond` field can be populated just above in the initialization of the `cesController` variable, like other fields.

File: operator/pkg/ciliumendpointslice/endpointslice.go [hunk 21]
```diff
@@ -59,8 +60,14 @@ func (c *Controller) initializeQueue() {
 		logfields.WorkQueueBurstLimit:  c.rateLimit.current.Burst,
 		logfields.WorkQueueSyncBackOff: defaultSyncBackOff,
 	}).Info("CES controller workqueue configuration")
-	c.queue = workqueue.NewRateLimitingQueueWithConfig(
-		workqueue.NewItemExponentialFailureRateLimiter(defaultSyncBackOff, maxSyncBackOff),
+
+	// single rateLimiter controls the number of failures in both queues
```
Comment (dlapcevic) [inline operator/pkg/ciliumendpointslice/endpointslice.go]: nit: Format the comment (first letter capital and end the comment with a period).

Also, this rate limiter is not only controlling the number failures. It's an actual rate limiter for the total number of processed events. used in `rateLimitProcessing()`.

File: operator/pkg/ciliumendpointslice/controller.go [hunk 22]
```diff
@@ -97,14 +118,23 @@ func registerController(p params) error {
...
 	p.Lifecycle.Append(cesController)
+	p.Job.Add(
```
Comment (dlapcevic) [inline operator/pkg/ciliumendpointslice/controller.go]: Should we call this here, or in the controller Start (in endpointslice.go), with the rest of the processes (e.g. "proc-queues")?

cc @ovidiutirla

File: operator/pkg/ciliumendpointslice/namespace.go [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+	slimcorev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+const (
+	priorityNamespaceAnnotation = "cilium.io/ces-namespace"
+)
+
+func (c *Controller) processNamespaceEvents(ctx context.Context) error {
+
```
Comment (dlapcevic) [inline operator/pkg/ciliumendpointslice/namespace.go]: nit: Remove the empty line at the beginning.

Same below for `onNamespaceUpstert()` and `onNamespaceDelete()` (also last line is empty here).

**Review by dlapcevic — APPROVED** [review]
LGTM

I only had a few nits.

The only remaining thing that is useful to consider is how we measure the benefits of this feature. It would be good to have metrics that show how many items were processed in the fast queue and what's the wait latency in the queue compared to the standard queue.

File: operator/pkg/ciliumendpointslice/controller.go (hunk 22 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumendpointslice/controller.go]: adding to job group is a safe operation, not important where we do that.

From code perspective, we are missing the init phase but it should still work fine even if we update the priority for the CES after some time (there is a chance the first reconcile will go on the standard queue until the namespace set is populated).

From readability, I think we should move this in endpointslice.go#149 (near the workerpool) and we should probably replace the workerpool with job.OneShots in a later PR.

Comment (dlapcevic) [reply operator/pkg/ciliumendpointslice/controller.go]: Sounds good.

**Comment by dlapcevic** [comment]
cc @marseel

File: operator/pkg/ciliumendpointslice/namespace.go [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+	"context"
+
+	k8s "github.com/cilium/cilium/pkg/annotation"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slimcorev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+const (
+	priorityNamespaceAnnotation = "cilium.io/ces-namespace"
```
Comment (marseel) [inline operator/pkg/ciliumendpointslice/namespace.go:16]: Let's document it in CES documentation.

**Review by marseel — CHANGES_REQUESTED** [review]
Very nice improvement, thanks!
I will review more in-depth later this week.

I'm wondering though, would it make sense to have a kube-system by default as a priority namespace?

**Comment by pchaigno** [comment]
This CI is failing because this branch is almost one month behind `main`. In general, whenever you are pushing an updated branch, it's always in your best interest to rebase to the latest `main`. Any push will invalidate previous CI results, so you might as well update to latest `main` to reduce the chances of having to push again because the CI fails.

File: operator/pkg/ciliumendpointslice/namespace.go (hunk 24 above)
Comment (dlapcevic) [reply operator/pkg/ciliumendpointslice/namespace.go:16]: Docs should be updated here https://github.com/dlapcevic/cilium/blob/main/Documentation/operations/upgrade.rst#117-upgrade-notes?

Anywhere else?

Comment (marseel) [reply operator/pkg/ciliumendpointslice/namespace.go:16]: I think most important is here: https://docs.cilium.io/en/latest/network/kubernetes/ciliumendpointslice/

File: operator/pkg/ciliumendpointslice/ceptocesmap.go [hunk 25]
```diff
@@ -167,6 +167,10 @@ func NewCESName(name string) CESName {
 	return CESName(resource.Key{Name: name})
 }
 
+func NewCESNameNamespace(name string, namespace string) CESName {
```
Comment (thorn3r) [inline operator/pkg/ciliumendpointslice/ceptocesmap.go]:
```suggestion
func NewCESNamespacedName(name string, namespace string) CESName {
```
small nit based on the convention used in apimachinery and controller-runtime

Also, should its return be a new type to distinguish from CESName objects that don't have a namespace? Do we need to keep both or can we just use namespaced CES now instead?

Comment (Kaczyniec) [reply operator/pkg/ciliumendpointslice/ceptocesmap.go]: Thanks, I changed the name. There are still places, where NewCESName is used and namespace information is not needed (I believe that's the case of cesStore in reconciler). I did not modify it since it didn't concern the double queue.

File: operator/pkg/ciliumendpointslice/namespace.go (hunk 24 above)
Comment (Kaczyniec) [reply operator/pkg/ciliumendpointslice/namespace.go:16]: I modified added a point to Configuration Options in ciliumendpointslice documentation

**Comment by pchaigno** [comment]
@Kaczyniec This branch is (exactly!) one month behind `main`. As a result the builds are failing. I'd recommend to take each push as an opportunity to rebase onto latest `main`.

File: Documentation/network/kubernetes/ciliumendpointslice.rst [hunk 26]
```diff
@@ -104,6 +104,9 @@ Several options are available to adjust the performance and behavior of the CES
 
 * You can also fine-tune rate-limiting settings for the Operator communications with the API-server. Refer to the ``--ces-*`` flags for the ``cilium-operator`` binary.
 
+* You can annotate priority namespaces by setting attribute ``cilium.io/ces-namespace`` to the value “priority”. When dealing with large clusters, the propagation of changes during Network Policy updates can be significantly delayed.
```
Comment (ghost) [inline Documentation/network/kubernetes/ciliumendpointslice.rst]: I'd say `annotation` instead of `attribute`, while it might seem obvious you mean an annotation, it might still be confusing for users less familiar with kubernetes.

**Review by ghost — APPROVED** [review]
Left one non-blocking comment, but docs good

File: operator/pkg/ciliumendpointslice/controller.go [hunk 27]
```diff
@@ -97,13 +118,17 @@ func registerController(p params) error {
...
 		ciliumEndpointSlice: p.CiliumEndpointSlice,
 		ciliumNodes:         p.CiliumNodes,
+		namespace:           p.Namespace,
 		slicingMode:         p.Cfg.CESSlicingMode,
 		maxCEPsInCES:        p.Cfg.CESMaxCEPsInCES,
 		rateLimit:           rateLimitConfig,
 		enqueuedAt:          make(map[CESName]time.Time),
 		metrics:             p.Metrics,
+		defaultCESSyncTime:  DefaultCESSyncTime,
+		priorityNamespaces:  make(map[string]struct{}),
```
Comment (marseel) [inline operator/pkg/ciliumendpointslice/controller.go:129]: blocking comment:
`priorityNamespaces` is accessed by two different goroutines:
- one in addToQueue - read
- one in processNamespaceEvents - write

This is a concurrent read/write from the map as far as I can see,  which can panic and should be protected by mutex.

File: operator/pkg/ciliumendpointslice/endpointslice.go [hunk 28]
```diff
@@ -78,11 +85,23 @@ func (c *Controller) onEndpointDelete(cep *cilium_api_v2.CiliumEndpoint) {
...
+}
+
+func (c *Controller) addToQueue(ces CESName) {
+	time.AfterFunc(c.defaultCESSyncTime, func() {
+		if _, exists := c.priorityNamespaces[ces.Namespace]; exists {
+			c.fastQueue.Add(ces)
+		} else {
+			c.standardQueue.Add(ces)
+		}
+		c.cond.Signal()
```
Comment (marseel) [inline operator/pkg/ciliumendpointslice/endpointslice.go:107]: Blocking comment:
We should lock here c.cond.L before adding element to one of queues, imagine the following scenario:
- Goroutine A get's lock in getQueue
- Goroutine B adds CES to queue here, but doesn't signal yet
- Goroutine A processes that CES, releases lock, and tries to get another one through getQueue
- As there is none of CES waiting in either of queues, it waits on condition
- Goroutine B signals and wakes up that goroutine even though there is no ces in the queues
- Goroutine A blocks on standardQueue, which doesn't have any element.

**Review by marseel — CHANGES_REQUESTED** [review]
I believe there are two race-conditions in the current implementation, let's address these before merging. 

I wonder if it would make sense to rethink concurrency here to make it a bit simpler to follow instead.

File: Documentation/network/kubernetes/ciliumendpointslice.rst (hunk 26 above)
Comment (Kaczyniec) [reply Documentation/network/kubernetes/ciliumendpointslice.rst]: Thanks, I changed it

File: operator/pkg/ciliumendpointslice/endpointslice.go (hunk 28 above)
Comment (Kaczyniec) [reply operator/pkg/ciliumendpointslice/endpointslice.go:107]: Thanks for pointing it out, I added these locks. We'll think further how can we make the concurrency clearer, this is the issue for it:
https://github.com/cilium/cilium/issues/34801

File: operator/pkg/ciliumendpointslice/controller.go [hunk 29]
```diff
@@ -63,15 +69,31 @@ type Controller struct {
...
+	rateLimiter workqueue.RateLimiter
 
 	enqueuedAt     map[CESName]time.Time
 	enqueuedAtLock lock.Mutex
 
 	wp *workerpool.WorkerPool
 
 	metrics *Metrics
+
+	defaultCESSyncTime time.Duration
```
Comment (marseel) [inline operator/pkg/ciliumendpointslice/controller.go]: nit: I would rename it to "SyncDelay" or "UpdateBatchPeriod" (similar to endpointslices in kcm).

File: operator/pkg/ciliumendpointslice/endpointslice.go [hunk 30]
```diff
@@ -59,8 +60,14 @@ func (c *Controller) initializeQueue() {
 		logfields.WorkQueueBurstLimit:  c.rateLimit.current.Burst,
 		logfields.WorkQueueSyncBackOff: defaultSyncBackOff,
 	}).Info("CES controller workqueue configuration")
-	c.queue = workqueue.NewRateLimitingQueueWithConfig(
-		workqueue.NewItemExponentialFailureRateLimiter(defaultSyncBackOff, maxSyncBackOff),
+
+	// Single rateLimiter controls the number of processed events in both queues.
+	c.rateLimiter = workqueue.NewItemExponentialFailureRateLimiter(defaultSyncBackOff, maxSyncBackOff)
+	c.fastQueue = workqueue.NewRateLimitingQueueWithConfig(
```
Comment (marseel) [inline operator/pkg/ciliumendpointslice/endpointslice.go]: nit: `NewRateLimitingQueueWithConfig` is deprecated. Please use `NewTypedRateLimitingQueueWithConfig` instead.

File: operator/pkg/ciliumendpointslice/endpointslice.go [hunk 31]
```diff
@@ -252,46 +288,70 @@ func (c *Controller) rateLimitProcessing() {
...
+		queue.Forget(key)
 		return
 	}
 
-	if c.queue.NumRequeues(key) < maxRetries {
-		c.queue.AddRateLimited(key)
+	if queue.NumRequeues(key) < maxRetries {
+		time.AfterFunc(c.rateLimiter.When(key), func() {
+			queue.Add(key)
+			c.cond.Signal()
```
Comment (marseel) [inline operator/pkg/ciliumendpointslice/endpointslice.go:361]: Shouldn't we lock here as well? Similar case to the previous example with race-condition.

File: operator/pkg/ciliumendpointslice/namespace.go [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+// onNamespaceUpsert modifies the Controller's list of priority namespaces if the namespace is modified.
+func (c *Controller) onNamespaceUpsert(ns *slimcorev1.Namespace) {
+	value, _ := k8s.Get(ns, priorityNamespaceAnnotation)
+	c.priorityNamespacesLock.Lock()
+	defer c.priorityNamespacesLock.Unlock()
+	if value == "priority" {
+		c.logger.Infof("Namespace has a priority annotation %s", ns.Name)
+		c.priorityNamespaces[ns.Name] = struct{}{}
+	} else {
+		c.logger.Infof("Namespace does not have priority: %s", ns.Name)
```
Comment (marseel) [inline operator/pkg/ciliumendpointslice/namespace.go]: Let's change it to debug level. During the bootstrap phase of the operator when it lists all namespaces, it would flood logs with all namespaces. 

We could print it however when this namespace was already priority, first by checking if this namespace is in `priorityNamespaces` map to indicate that namespace was removed from priority queue.

File: operator/pkg/ciliumendpointslice/endpointslice_test.go [hunk 33]
```diff
@@ -151,13 +156,195 @@ func TestIdentityModeSyncCESsInLocalCache(t *testing.T) {
...
+		}
+	}
+
+	for i := 0; i < 10; i++ {
+		cesController.processNextWorkItem()
+		if i < 4 {
+			standardQueueLen = 6
+			fastQueueLen = 3 - i
+		} else {
+			standardQueueLen = 9 - i
```
Comment (marseel) [inline operator/pkg/ciliumendpointslice/endpointslice_test.go]: nit: `6 - (i - 3)` will be a bit more readable

**Review by marseel — CHANGES_REQUESTED** [review]
Thanks for udpating PR, looks much better now :) 
 
I left two comments - one with a race-condition, one with the log flood.
Everything else marked as nit is non-blocking, but would be nice to fix if you can :)

File: operator/pkg/ciliumendpointslice/namespace.go (hunk 32 above)
Comment (Kaczyniec) [reply operator/pkg/ciliumendpointslice/namespace.go]: Now it's printed if the namespace was in the priority list and is deleted from it. Is it what you meant?

File: operator/pkg/ciliumendpointslice/endpointslice.go (hunk 31 above)
Comment (Kaczyniec) [reply operator/pkg/ciliumendpointslice/endpointslice.go:361]: Thanks, I corrected it

File: operator/pkg/ciliumendpointslice/controller.go [hunk 34]
```diff
@@ -63,15 +69,31 @@ type Controller struct {
...
+	rateLimiter workqueue.TypedRateLimiter[CESName]
 
 	enqueuedAt     map[CESName]time.Time
 	enqueuedAtLock lock.Mutex
 
 	wp *workerpool.WorkerPool
 
 	metrics *Metrics
+
+	SyncDelay time.Duration
```
Comment (marseel) [inline operator/pkg/ciliumendpointslice/controller.go]: nit: ah sorry for type when I was writing previous comment, should be `syncDelay`

File: operator/pkg/ciliumendpointslice/endpointslice.go [hunk 35]
```diff
@@ -47,7 +48,7 @@ const (
 
 	// Default CES Synctime, multiple consecutive syncs with k8s-apiserver are
 	// batched and synced together after a short delay.
-	DefaultCESSyncTime = 500 * time.Millisecond
+	SyncDelay = 500 * time.Millisecond
```
Comment (marseel) [inline operator/pkg/ciliumendpointslice/endpointslice.go]: nit: this can stay as "DefaultSyncTime".
I was trying to suggest only changing controller's field name.

File: operator/pkg/ciliumendpointslice/namespace.go [hunk 36]
```diff
@@ -0,0 +1,62 @@
...
+	if value == "priority" {
+		c.logger.Debugf("Namespace has a priority annotation %s", ns.Name)
+		c.priorityNamespaces[ns.Name] = struct{}{}
+	} else {
+		c.logger.Debugf("Namespace does not have priority: %s", ns.Name)
+		_, ok := c.priorityNamespaces[ns.Name]
+		if ok {
+			c.logger.Infof("Namespace %s removed from priority list.", ns.Name)
+		}
+		delete(c.priorityNamespaces, ns.Name)
```
Comment (marseel) [inline operator/pkg/ciliumendpointslice/namespace.go:50]: nit: you can move `delete` inside `if ok`

File: operator/pkg/ciliumendpointslice/ceptocesmap.go [hunk 37]
```diff
@@ -167,6 +167,10 @@ func NewCESName(name string) CESName {
 	return CESName(resource.Key{Name: name})
 }
 
+func NewCESNamespacedName(name string, namespace string) CESName {
```
Comment (marseel) [inline operator/pkg/ciliumendpointslice/ceptocesmap.go]: I believe this introduces regression. It will be probably easier to discuss on Slack, so I've started a thread: https://cilium.slack.com/archives/C069SNDMLA2/p1726061039032609

tldr: 
After this change, we will be inserting `CESName` with namespace information.
However,  `uniqueCESliceName` only checks by name without namespace information. So won't be able to detect conflicts of names of CES anymore. 

As now we have both `NewCESName` and `NewCESNamespacedName` - we should get rid of the old `NewCESName` to make sure we don't introduce regression somewhere. 
We could probably also get rid of `cesData` in `CESToCEPMapping` as it contains CESName -> Namespace mapping as it won't be anymore needed since CESName contains namespace information.

**Review by marseel — CHANGES_REQUESTED** [review]
One last comment related to `NewCESNamespacedName`. I've unpurposely ignored this part a bit in previous reviews :sweat_smile:

File: operator/pkg/ciliumendpointslice/namespace.go (hunk 32 above)
Comment (marseel) [reply operator/pkg/ciliumendpointslice/namespace.go]: yes, thanks!

File: operator/pkg/ciliumendpointslice/ceptocesmap.go (hunk 37 above)
Comment (Kaczyniec) [reply operator/pkg/ciliumendpointslice/ceptocesmap.go]: You're right, we changed it now to ensure that the information about namespace isn't taken into account where it isn't needed. resource.Key containing the namespace is now CESKey. CESName is the alias for string, used in contexts, where only the name of CES is needed (like in functions from ceptocesmap.go)

File: operator/pkg/ciliumendpointslice/manager.go [hunk 38]
```diff
@@ -98,17 +98,17 @@ func (c *cesMgr) createCES(name, ns string) CESName {
 	if name == "" {
 		name = uniqueCESliceName(c.mapping)
 	}
-	cesName := NewCESName(name)
+	cesName := CESName(name)
 	c.mapping.insertCES(cesName, ns)
```
Comment (marseel) [inline operator/pkg/ciliumendpointslice/manager.go:102]: nit: `c.mapping.insertCES` could use `CESKey` instead

File: operator/pkg/ciliumendpointslice/endpointslice.go [hunk 39]
```diff
@@ -147,21 +185,32 @@ func (c *Controller) Start(ctx cell.HookContext) error {
 		return err
 	}
 
+	c.Job.Add(
+		job.OneShot("proc-ns-events", func(ctx context.Context, health cell.Health) error {
+			return c.processNamespaceEvents(ctx)
```
Comment (marseel) [inline operator/pkg/ciliumendpointslice/endpointslice.go:190]: ~nit: we should block on the initial sync of namespaces

Imagine that the operator is starting and we have a priority namespace. We need to perform initial sync of namespaces, before adding CES updates to properly choose fast/slow queues. 

However, this is non-blocking, can be a follow up.

**Review by marseel — APPROVED** [review]
Thanks for the contribution! 
Looks good to me :smile:

File: operator/pkg/ciliumendpointslice/endpointslice.go (hunk 39 above)
Comment (marseel) [reply operator/pkg/ciliumendpointslice/endpointslice.go:190]: I've added that to https://github.com/cilium/cilium/issues/34801

Comment (dlapcevic) [reply operator/pkg/ciliumendpointslice/endpointslice.go:190]: Thanks for the suggestion. I updated the issue that tracks improvements for this feature: https://github.com/cilium/cilium/issues/34801
